### PR TITLE
Introduce immutables annotation processors

### PR DESCRIPTION
**Pull request by abdelhamidbakhta**
## PR description

This PR introduces [Immutables](https://immutables.github.io/) annotation processors to generate simple, safe and consistent value objects. 

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java [hunk 1]
```diff
@@ -14,169 +14,41 @@
  */
 package org.hyperledger.besu.ethereum.eth;
 
-import org.hyperledger.besu.util.number.PositiveNumber;
+import org.immutables.value.Value;
 
-import java.util.Objects;
+@Value.Immutable
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java:19]: so does this annotation mean that everything in this class is immutable

File: besu/src/test/java/org/hyperledger/besu/cli/options/EthProtocolOptionsTest.java [hunk 2]
```diff
@@ -116,24 +116,18 @@ public void parsesInvalidEwpMaxGetNodeDataOptionsShouldFail() {
...
-            PositiveNumber.fromInt(EthProtocolConfiguration.DEFAULT_MAX_GET_BLOCK_BODIES + 2))
-        .maxGetReceipts(
-            PositiveNumber.fromInt(EthProtocolConfiguration.DEFAULT_MAX_GET_RECEIPTS + 2))
-        .maxGetNodeData(
-            PositiveNumber.fromInt(EthProtocolConfiguration.DEFAULT_MAX_GET_NODE_DATA + 2))
-        .maxGetPooledTransactions(
-            PositiveNumber.fromInt(
-                EthProtocolConfiguration.DEFAULT_MAX_GET_POOLED_TRANSACTIONS + 2))
-        .eth65Enabled(!EthProtocolConfiguration.DEFAULT_ETH_65_ENABLED)
+    return ImmutableEthProtocolConfiguration.builder()
```
Comment (rain-on) [inline besu/src/test/java/org/hyperledger/besu/cli/options/EthProtocolOptionsTest.java:124]: q: when you cmd-click on "ImmutableEthProtocolConfiguration" - does it go to the interface? Or the generated class?
What about "builder()"?

File: besu/src/test/java/org/hyperledger/besu/cli/options/EthProtocolOptionsTest.java [hunk 3]
```diff
@@ -18,7 +18,7 @@
 import static org.mockito.Mockito.verifyNoInteractions;
 
 import org.hyperledger.besu.ethereum.eth.EthProtocolConfiguration;
-import org.hyperledger.besu.util.number.PositiveNumber;
+import org.hyperledger.besu.ethereum.eth.ImmutableEthProtocolConfiguration;
```
Comment (rain-on) [inline besu/src/test/java/org/hyperledger/besu/cli/options/EthProtocolOptionsTest.java:21]: do you need to run the annotation processor over EthProtocolConfiguration prior to writing this file? (i.e.to ensure ImmutableEthProtcolConfiguration exists)?

Does Auto-import etc. still work in the IDE?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java (hunk 1 above)
Comment (abdelhamidbakhta) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/EthProtocolConfiguration.java:19]: Yes, exactly

File: besu/src/test/java/org/hyperledger/besu/cli/options/EthProtocolOptionsTest.java (hunk 2 above)
Comment (abdelhamidbakhta) [reply besu/src/test/java/org/hyperledger/besu/cli/options/EthProtocolOptionsTest.java:124]: It goes to the generated class, in both cases.

File: besu/src/test/java/org/hyperledger/besu/cli/options/EthProtocolOptionsTest.java (hunk 3 above)
Comment (abdelhamidbakhta) [reply besu/src/test/java/org/hyperledger/besu/cli/options/EthProtocolOptionsTest.java:21]: You simply run your build or your tests and missing classes will be generated if needed. You might have an error the first time you run the build. But after that everything works fine, including auto import and all other features.

**Pull request closed without merging**